tensorflow/contrib/grid_rnn/python/kernel_tests/grid_rnn_test.py

```diff
@@ -0,0 +1,324 @@
+        self.assertAllClose(res[0], [[0.9032144, 0.9032144]])
+        self.assertAllClose(res[1], [[2.79966092, 2.79966092, 0.9032144, 0.9032144]])
+
+        g3, s3 = cell(None, m)
+        sess.run([tf.initialize_all_variables()])
+        res = sess.run([g3, s3], {m.name: res[1]})
+        self.assertEqual(res[0].shape, (1, 2))
+        self.assertEqual(res[1].shape, (1, 4))
+        self.assertAllClose(res[0], [[0.92727238, 0.92727238]])
+        self.assertAllClose(res[1], [[3.3529923, 3.3529923, 0.92727238, 0.92727238]])
+
+  """
+  3-LSTM
+  """
+  def testGrid3LSTMCell(self):
+    with self.test_session() as sess:
+      with tf.variable_scope('root', initializer=tf.constant_initializer(0.5)):
+        x = tf.zeros([1, 3])
+        m = tf.zeros([1, 12])
+        cell = tf.contrib.grid_rnn.Grid3LSTMCell(2, input_size=3, use_peepholes=True)
+        self.assertEqual(cell.output_size, 2)
+        self.assertEqual(cell.input_size, 3)
+        self.assertEqual(cell.state_size, 12)
+
+        g, s = cell(x, m)
+        sess.run([tf.initialize_all_variables()])
+        res = sess.run([g, s], {x.name: np.array([[1., 1., 1.]]),
+                                m.name: np.array([[0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, -0.1, -0.2, -0.3, -0.4]])})
+        self.assertEqual(res[0].shape, (1, 2))
+        self.assertEqual(res[1].shape, (1, 12))
+
+        self.assertAllClose(res[0], [[0.96892911, 0.96892911]])
+        self.assertAllClose(res[1], [[2.45227885, 2.45227885, 0.96892911, 0.96892911,
+                                      1.33592629, 1.4373529, 0.80867189, 0.83247656,
+                                      0.7317788, 0.63205892, 0.56548983, 0.50446129]])
+
+  """
+  Edge cases
+  """
+  def testGridRNNEdgeCasesLikeRelu(self):
+    with self.test_session() as sess:
+      with tf.variable_scope('root', initializer=tf.constant_initializer(0.5)):
+        x = tf.zeros([3, 2])
+        m = tf.zeros([0, 0])
+
+        # this is equivalent to relu
+        cell = tf.contrib.grid_rnn.GridRNNCell(num_units=2, num_dims=1, input_dims=0, output_dims=0,
+                                               non_recurrent_dims=0, non_recurrent_fn=tf.nn.relu)
+        g, s = cell(x, m)
+        sess.run([tf.initialize_all_variables()])
+        res = sess.run([g, s], {x.name: np.array([[1., -1.], [-2, 1], [2, -1]])})
+        self.assertEqual(res[0].shape, (3, 2))
+        self.assertEqual(res[1].shape, (0, 0))
+        self.assertAllClose(res[0], [[0, 0], [0, 0], [0.5, 0.5]])
+
+  def testGridRNNEdgeCasesNoOutput(self):
+    with self.test_session() as sess:
+      with tf.variable_scope('root', initializer=tf.constant_initializer(0.5)):
+        x =  tf.zeros([1, 2])
+        m = tf.zeros([1, 4])
+
+        # This cell produces no output
+        cell = tf.contrib.grid_rnn.GridRNNCell(num_units=2, num_dims=2, input_dims=0, output_dims=None,
+                                               non_recurrent_dims=0, non_recurrent_fn=tf.nn.relu)
```

Comment:
`input_dims` and `output_dims` are the dimension indices of the gridRNN cell. These are different from `input_size` and `output_size`.